[move-analyzer] function signature help

0xYao opened the pull request:
<!--
Thank you for sending a PR. We appreciate you spending time to help improve the Move Language.

The project is undergoing daily changes. Pull Requests will be reviewed and responded to as time permits.
-->

## Motivation

This is a follow-up PR from https://github.com/move-language/move/pull/338. Providing signature help should help developers when calling a function inside a file.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/move-language/move/blob/main/CONTRIBUTING.md#developer-workflow)?

Yes.

## Test Plan

(Share your test plan here. If you changed code, please provide us with clear instructions for verifying that your changes work.)

- Main test cases have been discussed and covered in `signatureHelp.test.ts`
- Open the extension host in debug mode. You should see a signature help popup when calling a function.
   - There are two key parts of a signature help. The first is the signature label and the second is the position of the active parameter, and you should see the expected signature label and the expected active parameter position when you move the cursor around.

## Demo

https://www.loom.com/share/16e5c241c78a4433bae71046089b73b6

awelc commented on language/move-compiler/src/parser/lexer.rs:
```diff
@@ -402,6 +414,34 @@ impl<'input> Lexer<'input> {
         self.cur_end = self.cur_start + len
     }
+
+    /// Construct the text info for every token in the lexer by advancing until EOF
```
It seems like this is used to parse a single line - is looking for EOF as the end of iteration a correct choice here(@tnowacki)?

0xYao replied:
@tnowacki a friendly reminder for this^

awelc commented on language/move-compiler/src/parser/lexer.rs:
```diff
@@ -82,6 +82,18 @@ pub enum Tok {
+    pub end: usize,
+}
+
+#[derive(Copy, Clone, Debug, Default)]
```
Please add documentation to these definitions.

0xYao replied:
done

awelc commented on language/move-analyzer/src/symbols.rs:
```diff
@@ -2190,6 +2198,228 @@ pub fn on_hover_request(context: &Context, request: &Request, symbols: &Symbols)
 }
 
+/**
+ * Returns a tuple where the first value is the index of the text info of the function name and the second value is the text info of the function
```
Please add explanation of "text info" for the function describes (function name and position/range of the name?)

awelc commented on language/move-analyzer/src/symbols.rs:
```diff
@@ -2190,6 +2198,228 @@ pub fn on_hover_request(context: &Context, request: &Request, symbols: &Symbols)
+}
+
+/**
+ * Find the position of the active parameter given a vector of parsed tokens
```
Please add more explanation of what parsed tokens contains (what kind of line was parsed to obtain the tokens).

0xYao replied:
done

awelc commented on language/move-analyzer/src/symbols.rs:
```diff
@@ -2190,6 +2198,228 @@ pub fn on_hover_request(context: &Context, request: &Request, symbols: &Symbols)
+    active_parameter
+}
+
+fn collect_parameter_info_from_ident(ident: &IdentType) -> Vec<ParameterInformation> {
```
Please dd documentation for the function.

0xYao replied:
done

awelc reviewed (changes requested):
This is a relatively complicated addition and I would really appreciate it if you could add a higher level description of how the algorithm works. I have a high-level idea here but it's difficult to piece together all the details from the code itself - a textual description of the algorithm (what is parsed and when and how the parsed information is used) would go a long way (it does not have to be super long).

Also, I have build and and installed both the language server and the extension from this branch, but things to do not work for me quite the same as in the demo. In particular, while I am getting a match (vi text highlighting) between parameter name (being typed) and parameter type (in the hover box), I do not get the parameter name auto-completion (see attached screenshot). I am using VSCode 1.70.1 and all tests pass with this version (after I modified runTests.rs to include it).

![image](https://user-images.githubusercontent.com/1724397/184987008-74a89858-7e50-4a61-9ddf-3267203c1d56.png)

awelc commented:
I have also discovered a problem with the previous PR (https://github.com/move-language/move/pull/338) underlying this one. I added a comment there even though the PR is already merged: https://github.com/move-language/move/pull/338/files#r948387534

Please provide a fix first along with appropriate test before the current PR is merged.

0xYao commented:
> I have addressed this in #389 (please note that I removed the `IdentType` for functions is now available via `Symbols::file_mods`.

Sounds good, let me push the fixes using the latest changes.

0xYao commented:
> I do not get the parameter name auto-completion

Sorry, what do you mean by parameter name auto-completion? Is this addressed in your #389 PR?

0xYao commented:
Given the cursor position of the current line, I need to find out the function name and the corresponding module that it belongs. It seems like I can reuse some of the logic from `Symbolicator::mod_call_symbols`, but maybe it's better to do custom string parsing and then extract the correct information from `file_mods`?

awelc commented:
> Given the cursor position of the current line, I need to find out the function name and the corresponding module that it belongs.

I thought that you could simply use `Symbols::file_mods`. In most cases there will be one module per file, and even if there isn't, iterating over the list of modules in the set shouldn't be a huge deal (in terms of speed and complexity).  Unless I misunderstand what information you need to obtain and how.

0xYao commented:
> I thought that you could simply use `Symbols::file_mods`. In most cases there will be one module per file, and even if there isn't, iterating over the list of modules in the set shouldn't be a huge deal (in terms of speed and complexity). Unless I misunderstand what information you need to obtain and how.

Additionally, I also need to find the current enclosing scope given a cursor position, and determine which module the developer is calling the function from.

awelc commented:
> Additionally, I also need to find the current enclosing scope given a cursor position, and determine which module the developer is calling the function from.

Good point. Here are some thoughts. 

First off, it may make sense to special-case implementation for one-module-per-file - we don't have to worry about multiple modules in this case (which should be the common case).

In terms of figuring out module scope (in terms of file lines), when constructing `ModuleDefs`, we iterate over typed AST unique map that gives us location of modules (I think) in `pos`:
https://github.com/move-language/move/blob/70b34a66473c34ad30d101290b249f2db3c847a2/language/move-analyzer/src/symbols.rs#L708

We currently only store the starting position of the module but this should give us the ending position as well, which we could use to figure out module's scope within a file. To make it fully robust we'd probably have to make an adjustment for a new line (or lines) being inserted that would change start/end of modules in the file, but I am not sure if this is feasible at the moment. 

If this does not work for some reason, then I believe a function call can only be inserted inside another function. If we then remembered both starting line of function definition (we already have this in `FunctionDef`) and the ending line (we could conceivably obtain it from the outermost code block in function's body) we could use `Symbol::file_mods` to iterate over all functions in a given file and figure out if the cursor is located inside one of them. This is potentially more fragile than the previous suggestion, but I haven't checked if module start/end locations are indeed faithfully represented in the AST.

0xYao commented on language/move-analyzer/src/symbols.rs:
```diff
@@ -2190,6 +2198,228 @@ pub fn on_hover_request(context: &Context, request: &Request, symbols: &Symbols)
 }
 
+/**
+ * Returns a tuple where the first value is the index of the text info of the function name and the second value is the text info of the function
```
done

0xYao commented on language/move-analyzer/src/symbols.rs:
```diff
@@ -2190,6 +2198,228 @@ pub fn on_hover_request(context: &Context, request: &Request, symbols: &Symbols)
+/**
+ * Returns a tuple where the first value is the index of the text info of the function name and the second value is the text info of the function
+ */
+fn find_closet_fn_in_text_info_vec<'a>(
```
fixed

0xYao commented on language/move-analyzer/src/symbols.rs:
```diff
@@ -2190,6 +2198,228 @@ pub fn on_hover_request(context: &Context, request: &Request, symbols: &Symbols)
+                if parenthesis_delta == 0 {
+                    // The identifier to the left of the left parenthesis is the function.
+                    match text_info_vec.get(i - 1) {
+                        Some(text_info) => {
```
fixed

0xYao commented:
> First off, it may make sense to special-case implementation for one-module-per-file - we don't have to worry about multiple modules in this case (which should be the common case).

I am going with the case where we have one-module-per-file for now. I have added a note in the `on_signature_help_request` function in case we want to address this in the future.

awelc commented:
> I am going with the case where we have one-module-per-file for now. I have added a note in the `on_signature_help_request` function in case we want to address this in the future.

What are you going to to in the short term then if we have two modules in a file and each module defines a function with the same name? Are you going to list them both in the function signature help?

0xYao commented:
> What are you going to to in the short term then if we have two modules in a file and each module defines a function with the same name? Are you going to list them both in the function signature help?

The current implementation just returns the signature help of the function in the first module with that function name that we encounter in `file_mods`. However, if this behaviour is not acceptable, I can update the implementation to return the signature help of the function inside the enclosing module only in the short-term.

But feel free to propose an alternative solution.

0xYao commented:
@awelc can you please review?

awelc commented on language/move-analyzer/src/symbols.rs:
```diff
@@ -2206,6 +2210,288 @@ pub fn on_hover_request(context: &Context, request: &Request, symbols: &Symbols)
+///
+/// The text info result includes the name and the range of the function in a given line.
+///
+/// A non-zero parenthesis_delta value means the current index is in a nested function scope.
```
Please move the comments concerning algorithmic details for all functions (which are very useful, by the way!) to the beginning of the function itself, leaving the documentation comments more succinct.

awelc commented on language/move-analyzer/src/symbols.rs:
```diff
@@ -2206,6 +2210,288 @@ pub fn on_hover_request(context: &Context, request: &Request, symbols: &Symbols)
+///
+/// A non-zero parenthesis_delta value means the current index is in a nested function scope.
+/// For example, consider `fun_1(a, fun_2(b, c, d), e)` and start_idx is at "e", which means
+/// we want to find out which function "e" belongs to. Since we are interating backwards from
```
```suggestion
/// we want to find out which function "e" belongs to. Since we are iterating backwards from
```

awelc reviewed (changes requested):
Thank you for the updates! The newly added comments helped a lot in understanding what's going on!

I still have a few questions/concerns:
- as I indicated before, I cannot reproduce the behavior shown in the video clip attached to the PR - the auto-completion for parameter names does not work for me (is there a special setting that I need to toggle?)
- I wonder if the current version of the algorithm works if a function signature is spread across different lines, and also if the developer wants to put arguments on different lines; I would say that we really need the former while the latter would be nice to have but could be delegated to future work
- I would much prefer to have the multi-module-in-one-file case handled by providing multiple functions in the signature help if multiple of them are a match (e.g. have sufficient number of parameters); as it is now, the risk is that only the "wrong" function signature will be provided which would be quite confusing to the developer
